Change the new board modal to be selected in tile mode in collaboration with task number [33864]

bsatarnejad opened the pull request:
Change the new board modal that user can select board types in tile mode.

bsatarnejad committed:
Change the new board modal to be selected in tile mode

openprojectci commented on frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts:
```diff
@@ -40,7 +40,8 @@ import {LoadingIndicatorService} from "core-app/modules/common/loading-indicator
 import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/2ae7cbab6a846f63f4c17f118ded9c3201310783/frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts#L41"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

bsatarnejad committed:
change the test of selecting a board

oliverguenther commented:
![Screenshot from 2020-07-30 09-06-00](https://user-images.githubusercontent.com/459462/88891712-e8499680-d243-11ea-9ec4-abc4280cd7d0.png)

Hi @bsatarnejad , I have some suggestions for icons that we could use to differentiate between the board types:

- Status is connected to a workflow: `icon-workflow`
- Assignee can use `icon-user`
- For versions we don't really have a dedicated icon, but `icon-getting-started` is this arrow of increasing numbers which might work fine.

Also, the modal has no heading, we should probably add one. 

I was first clicking on one of the buttons and thought I would get some information about what they are. I then found the hover tooltip, but that might be too late for users to note. Could we make it so that clicking the button will show a text below it and enable a highlight "Create" button at the bottom of the footer?

Could use more feedback on that as well. @lindenthal @wielinde ?

wielinde commented:
Hey, here my quick feedback:

- I think the icons are optimized for smaller font sizes. So they work for now but don't look great.
- I don't like the workflow icon in general, so not for workflows. That icon for me is a sync icon. Not a workflow. Some goes for the automatic scheduling icon. I think it confuses more than what it helps. Some things are hard do describe in icons. Even in words it is hard. What does "Action board" mean? Why don't we call them "Status board", "Versions board", "Assignee board" etc. and skip the term "Action board"?
- I agree, one sentence per board, that is always visible, would be a great help for the user.
- The tiles in the screenshot don't look like anything that we have. It's not clear that they are clickable or simply info boxes. So probably we need to add some new paradigm to your style guide. And that is "Cards". I would simply have a peek on how Google Material is solving that and copy where we can. They use cards a lot and somehow they manage to make it clear that they are clickable. If I remember right, we have similar challanges for the openproject.org website. I am not sure if we solved them there.
- Yes, the header is missing. Maybe a call to action like "Please chose the type of board you need" or something like that

bsatarnejad committed:
changing the style of tiles

openprojectci commented on frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts:
```diff
@@ -40,7 +40,8 @@ import {LoadingIndicatorService} from "core-app/modules/common/loading-indicator
 import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/fc549e95616a70495e3ba1ad91a26b9a6621fbaf/frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts#L40"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther commented:
Hi @bsatarnejad 

functionality-wise, this is already much better than the behavior before. It's good that we show all the text.

I like it better when the color of the icon/heading is identical:
![Screenshot from 2020-08-03 14-51-26](https://user-images.githubusercontent.com/459462/89184321-e4e24200-d598-11ea-9e68-53d43f520a64.png)

Also a few minor things:

1. The modal has an inner scroll behavior. Can we get rid of this? It appears only to be a few pixels
2. The modal does not have a header but a header underline. I'd say we add a heading
3. The border color/background is a bit too strong I think, but not as important
4.  I'm not too fond of the + button. What about making the entire tile clickable with a pointer cursor? I do see your point with adding an explicit button to show that this can be clicked. Not too sure with this one

![2020-08-03-14-50-25](https://user-images.githubusercontent.com/459462/89184325-e6136f00-d598-11ea-90c6-977565edba3d.png)

Also, to make further changes more simple, could you make this Tile/Grid into its own component with styles? This will make it easier to encapsulate the styles and simply pass an array of options as an input. What do you think?

Best
Oliver

bsatarnejad commented:
Hi @oliverguenther .Thanks for your feedback.
1. yes i solved this.
2. In release 10.6, we add a header to this modal, but now that we are in dev branch, it is not shown.
3. Ok i will make it better.
4. I agree with you, i will make the whole tile clickable.
5. I think if we need this kind of tiles in the future, it would be worthy to have it as a component. I think it may not take too much time.

bsatarnejad committed:
create a component for tile-view

bsatarnejad committed:
a test for the tile-view component

bsatarnejad committed:
remove additional f from test

oliverguenther committed:
Merge branch 'dev' into improve-new-board-modal

openprojectci commented on frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts:
```diff
@@ -37,9 +37,11 @@ import {BoardService} from "core-app/modules/boards/board/board.service";
 import {HalResourceNotificationService} from "core-app/modules/hal/services/hal-resource-notification.service";
+import { ITileViewEntry } from '../tile-view/tile-view.component';
+
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/a9c021538e352a81028c1680e97b16dea2976a8f/frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts#L42"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther commented on frontend/src/app/modules/boards/new-board-modal/new-board-modal.component.ts:
```diff
@@ -101,5 +120,37 @@ export class NewBoardModalComponent extends OpModalComponent {
+  }
+  private addIcon(tiles:ITileViewEntry[]) {
+    tiles.forEach(element => {
+      if (element.attribute === 'assignee') {
```
I would suggest to move these icons into the *ActionService classes themselves. Add a property in the baseclass (similar to filterName here: https://github.com/opf/openproject/blob/dev/frontend/src/app/modules/boards/board/board-actions/board-action.service.ts#L43) This would allow to keep the definition of name/description and icon in the service? What do you think?

oliverguenther commented on frontend/src/app/modules/boards/tile-view/tile-view.component.sass:
```diff
@@ -0,0 +1,36 @@
+  grid-auto-rows: 200px
+  grid-column-gap: 10px
+  grid-row-gap: 10px
+  min-width: 405px;
```
There are a few invalid semicolon in this sass file

bsatarnejad committed:
making changes in styles

bsatarnejad committed:
Changing methods of adding icons, text and description to tiles

oliverguenther committed:
Some minor styling changes

- Use primary-color-dark so the color matches the theme
- Use flex for positioning of icon + label
- Remove unused / unnecessary styles

oliverguenther reviewed (approved):
Hi @bsatarnejad , great job! I pushed a few minor changes to the tiling styles which change how the icon + text are selected, and also use the primary dark color variable (so the colors match the users' theme). 

Please take a look at them and if you agree with these changes, the PR is good to merge from my side once travis agrees. There is still a few tests failing because they look for an uppercase text (Status, Assignee etc.).

bsatarnejad committed:
correct the errors in checks

bsatarnejad committed:
solving Travis check issues

oliverguenther committed:
Ensure test checks for title case
